RecoEgamma/EgammaIsolationAlgos/interface/EleTkIsolFromCands.h

```diff
@@ -65,50 +58,84 @@ class EleTkIsolFromCands {
+      : cfg_{cfg}, tracks_{&tracks} {}
+  explicit EleTkIsolFromCands(Configuration const& cfg,
+                              pat::PackedCandidateCollection const& cands,
+                              PIDVeto pidVeto = PIDVeto::NONE)
+      : cfg_{cfg}, cands_{&cands}, pidVeto_{pidVeto} {}
 
-  std::pair<int, double> calIsol(const reco::TrackBase& trk,
-                                 const pat::PackedCandidateCollection& cands,
-                                 const PIDVeto = PIDVeto::NONE) const;
-  std::pair<int, double> calIsol(const double eleEta,
-                                 const double elePhi,
-                                 const double eleVZ,
-                                 const pat::PackedCandidateCollection& cands,
-                                 const PIDVeto = PIDVeto::NONE) const;
-
-  std::pair<int, double> calIsol(const reco::TrackBase& trk, const reco::TrackCollection& tracks) const;
-  std::pair<int, double> calIsol(const double eleEta,
-                                 const double elePhi,
-                                 const double eleVZ,
-                                 const reco::TrackCollection& tracks) const;
-
-  //little helper function for the four calIsol functions for it to directly return the pt
-  template <typename... Args>
-  double calIsolPt(Args&&... args) const {
-    return calIsol(std::forward<Args>(args)...).second;
-  }
+  static edm::ParameterSetDescription pSetDescript();
 
   static PIDVeto pidVetoFromStr(const std::string& vetoStr);
-  static bool passPIDVeto(const int pdgId, const EleTkIsolFromCands::PIDVeto pidVeto);
+
+  struct Output {
+    const int nTracks;
+    const double ptSum;
+  };
+
+  Output operator()(const reco::TrackBase& electronTrack);
 
 private:
-  static bool passTrkSel(const reco::TrackBase& trk,
-                         const double trkPt,
-                         const TrkCuts& cuts,
-                         const double eleEta,
-                         const double elePhi,
-                         const double eleVZ);
+  // For each electron, we want to try out which tracks are in a cone around
+  // it. However, this will get expensive if there are many electrons and
+  // tracks (Phase II conditions). In particular, calling
+  // reco::TrackBase::eta() many times is costy because eta is not precomputed.
+  // To solve this, we first cache the tracks in a simpler data structure in
+  // which eta is already computed (SimpleTrack). Furthermore, the tracks are
+  // preselected by the cuts that can already be applied without considering
+  // the electron. Note that this has to be done twice, because the required
+  // preselection is different for barrel and endcap electrons.
+  //
+  class SimpleTrack {
+  public:
+    SimpleTrack(reco::TrackBase const& trk) : pt{trk.pt()}, eta{trk.eta()}, phi{trk.phi()}, vz{trk.vz()} {}
+    const double pt;
+    const double eta;
+    const double phi;
+    const double vz;
+  };
+
```

Comment:
Since this is a pr aiming to performance, a comment which maybe is naive: isn't this an opportunity to introduce a simple SOA and a way to access it through some kind of view (e.g. a struct with references as data members)? But I might be reading the code incorrectly..

Reply:
This is definitely an opportunity to introduce this. Such a simple track-SOA could then also be reused when we optimize the other worst offenders in RecoEgamma at a later point (KF track matching and conversion finding), since the underlying problem is basically the same there.

Just to avoid reinventing the wheel here: do you know of existing CMSSW code that does what you describe, so I could use it as inspiration or in the best case even reuse it? I also found the [FWCore/SOA](https://github.com/cms-sw/cmssw/tree/master/FWCore/SOA) package quite fitting. I think it could be nicely applied here.

Reply:
I ping @Dr15Jones, since he is the person which invented the FWCore/SOA. I think it makes sense in this context.

Reply:
`edm::soa::Table` and `edm::SoATuple` are both generic implementations of SoA. `Table` was designed to work well as an Event data product (in theory could handle being made persistent with a bit of work). `SoATuple` is really meant for internal use and does better memory packing.

Reply:
Hi @Dr15Jones thanks for your answer! So with "internal use" you mean use inside the framework implementation details? Does that mean I made the right pick with `edm::soa::Table`?

Reply:
I actually meant internal details of a module or the framework but not really meant to be used as an data product. `edm::soa::Table` will work either as an internal detail or as something to pass between modules via the Event.